rfd/0124-plugin-secrets.md

```diff
@@ -0,0 +1,175 @@
+---
+authors: Michael Wilson (michael.wilson@goteleport.com)
+state: draft
+---
+
+# RFD 124 - Plugin static credentials
+
+### Required Approvers
+
+* Engineering @r0mant, @justinas
+* Security @reed
+* Product: @klizhentas
+
+## What
+
+Allow storage of plugin static credentials within Teleport in a safe and secure way.
+
+## Why
+
+As we grow our plugin support, we have discovered that the various services we
+hope to integrate do not offer a uniform mechanism for configuration. For instance,
+Slack uses oauth2, but Okta requires an API token, and Jamf requires a username and
+password.
+
+Today, the plugin houses a credentials object that allows for this sort of storage.
+However, unlike an oauth authorization code, static credentials pose a much greater risk of
+exposure, and rotating credentials can be a time consuming process.  To reduce the risks associated
+with storing static credentials within Teleport and to allow for easier credentials rotation,
+we would like to create a new resource within Teleport that is handled in an explicitly secure
+way.
+
+## Details
+
+### The `PluginStaticCredentials` object
+
+A new object will be needed to house static credentials. At present it will support:
+
+- API tokens
+- Usernames and passwords
+- OAuth client ID and secret
+
+This object has very strict access control requirements that will be explained in detail
+in the **Access Control** section. On creation, the plugin and static credentials will
+receive a UUID called the `plugin-label`. The static credentials will have an internal
+label that receives this value. Additionally, one plugin may have _many_ credentials
+for the purposes of credentials rotation.
+
+```yaml
+kind: plugin_static_credentials
+version: v1
+metadata:
+  name: credentials-name
+  labels:
+    # This is a unique plugin label generated randomly at plugin creation.
+    # It cannot be modified by the user.
+    teleport.internal/plugin: 48398071-9860-4be6-9f29-ca6ca5bc7155
+    # additional labels to uniquely identify the credentials.
+    label1: value1
+    type: slack
+spec:
+  credentials:
+    # Only one of these credential types must be defined at a time. All values in this
+    # are string literals and not file locations.
+    api_token: example-token
```

Comment:
Can we note if we'll be reusing the existing `Plugin*Credentials` types for variants here?

Reply:
I was thinking not so as to avoid compatibility concerns. I'll note it.